### PR TITLE
Master thread safe loader

### PR DESCRIPTION
**Pull request by Akm0d**
### What does this PR do?
Merge PR from develop https://github.com/saltstack/salt/pull/50655 which was a new version of #39948, #48598, and #50648 so the description has been copied from there. This PR will hopefully resolve the problems that existed in the original PRs. The code is identical to #50648, but this PR is based on the develop branch instead of the 2018.3.3 branch because Jenkins obviously cannot run tests for branches based on the 2018.3.3 branch.

There was a race condition in the salt loader when injecting global values (e.g. "pillar" or "salt") into modules. One effect of this race condition was that in a setup with multiple threads, some threads may see pillar data intended for other threads or the pillar data seen by a thread might even change spuriously.

There have been earlier attempts to fix this problem (#27937, #29397). These patches tried to fix the problem by storing the dictionary that keeps the relevant data in a thread-local variable and referencing this thread-local variable from the variables that are injected into the modules.

These patches did not fix the problem completely because they only work when a module is loaded through a single loader instance only. When there is more than one loader, there is more than one thread-local variable and the variable injected into a module is changed to point to another thread-local variable when the module is loaded again. Thus, the problem resurfaced while working on #39670.

This patch attempts to solve the problem from a slightly different angle, complementing the earlier patches: The value injected into the modules now is a proxy that internally uses a thread-local variable to decide to which object it points. This means that when loading a module
again through a different loader (possibly passing different pillar data), the data is actually only changed in the thread in which the loader is used. Other threads are not affected by such a change.

This means that it will work correctly in the current situation where loaders are possibly created by many different modules and these modules do not necessary know in which context they are executed. Thus it is much more flexible and reliable than the more explicit approach
used by the two earlier patches.
What issues does this PR fix or reference?

This PR fixes problems that surfaced when developing the parallel runners feature (#39670), but is also related to problems reported earlier (#23373). The problems described in #29028 might also be related, though this needs further confirmation.
Previous Behavior

Changes to pillar data in one thread might have influenced pillar data in a different thread. Whether or when this problem actually appeared was unpredictable (due to the nature of a race condition).
New Behavior

Changes to pillar data in one thread will never affect the pillar data in a different thread.
Tests written?

Yes
Regression Potential

The change to the loader code itself is rather small, but it affects a central component. So if there is a problem, the impact might affect nearly all components.

### CODE REVIEW

**Comment by max-arnold** [comment]
@Akm0d FYI: the backported PR also needs #51499 to fix the infinite recursion bug.

**Comment by Akm0d** [comment]
I'm getting this error on all the tests now: ```Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/integration/output/test_output.py", line 97, in test_output_yaml_namespaced_dict_wrapper
    assert isinstance(loaded_yaml, dict)
AssertionError```

**Comment by Akm0d** [comment]
It looks like the PR fixing https://github.com/saltstack/salt/issues/49269 was merged into 2017 and 2018 but not into 2019.2.1

**Comment by Akm0d** [comment]
The threadsafe loader has trouble representing undefined data types in the yamldumper (similar to https://github.com/saltstack/salt/issues/43694).

**Comment by Akm0d** [comment]
re-run full opensuse15

**Comment by Akm0d** [comment]
re-run full centos7

**Review by garethgreenaway — DISMISSED** [review]

**Review by s0undt3ch — APPROVED** [review]
There's more than the thread proxy stuff right?
Separate changes?

**Comment by Akm0d** [comment]
re-run full macosxmojave

**Comment by Akm0d** [comment]
re-run full all

**Comment by Akm0d** [comment]
re-run full macosxmojave

**Comment by Akm0d** [comment]
re-run full debian10

**Comment by Akm0d** [comment]
re-run full opensuse15

**Comment by Akm0d** [comment]
re-run full windows2019

**Comment by Akm0d** [comment]
re-run full macosxmojave

**Comment by Akm0d** [comment]
re-run full windows2019

**Comment by Akm0d** [comment]
> There's more than the thread proxy stuff right?
> Separate changes?

This is all the threadsafe loader + other changes required to make tests pass that failed with the threadsafe loader

**Comment by Akm0d** [comment]
re-run full macosxmojave

**Comment by Akm0d** [comment]
On this PR build the `Py3/macosxmojave` build hangs right after `test_unsupported_kwarg`. On a PR where the same build is passing, the subsequent test gives the following log:
```[2019-10-31T22:51:04.911Z]          test_unsupported_kwarg (integration.shell.test_arguments.ArgumentTestCase) ... OK (0.283s)
[2019-10-31T22:51:56.696Z]          test_pam_auth_invalid_user (integration.shell.test_auth.AuthTest) ... 16:51:55,145 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:51:56.696Z]        16:51:55,146 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:51:56.696Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:51:56.696Z]        16:51:55,146 [ERROR   ] retcode: 54
[2019-10-31T22:51:56.696Z]        16:51:55,178 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:51:56.696Z]        16:51:55,179 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:51:56.696Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:51:56.696Z]        16:51:55,179 [ERROR   ] retcode: 54
[2019-10-31T22:51:56.696Z]        16:51:55,210 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:51:56.696Z]        16:51:55,211 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:51:56.696Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:51:56.696Z]        16:51:55,211 [ERROR   ] retcode: 54
[2019-10-31T22:51:56.696Z]        16:51:55,243 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:51:56.696Z]        16:51:55,243 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:51:56.696Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:51:56.696Z]        16:51:55,244 [ERROR   ] retcode: 54
[2019-10-31T22:52:02.974Z]        16:52:01,936 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:52:02.974Z]        16:52:01,936 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:52:02.974Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:52:02.974Z]        16:52:01,937 [ERROR   ] retcode: 54
[2019-10-31T22:52:02.974Z]        16:52:01,971 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:52:02.974Z]        16:52:01,971 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:52:02.974Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:52:02.974Z]        16:52:01,971 [ERROR   ] retcode: 54
[2019-10-31T22:52:02.974Z]        16:52:02,007 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:52:02.974Z]        16:52:02,007 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:52:02.974Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:52:02.974Z]        16:52:02,008 [ERROR   ] retcode: 54
[2019-10-31T22:52:02.974Z]        16:52:02,077 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:52:02.974Z]        16:52:02,078 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:52:02.974Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:52:02.974Z]        16:52:02,078 [ERROR   ] retcode: 54
[2019-10-31T22:52:09.109Z]        OK (64.618s)
[2019-10-31T22:53:11.101Z]          test_pam_auth_valid_group (integration.shell.test_auth.AuthTest) ... 16:53:03,447 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:53:11.101Z]        16:53:03,447 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.101Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.101Z]        16:53:03,447 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.101Z]        16:53:03,479 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:03,480 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:03,480 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.102Z]        16:53:03,512 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:03,512 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:03,513 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.102Z]        16:53:03,549 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:03,549 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:03,550 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.102Z]        16:53:10,109 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:10,110 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:10,110 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.102Z]        16:53:10,145 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:10,145 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:10,146 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.102Z]        16:53:10,181 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:10,181 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:10,181 [ERROR   ] retcode: 54
[2019-10-31T22:53:11.102Z]        16:53:10,248 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:53:11.102Z]        16:53:10,249 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:53:11.102Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:53:11.102Z]        16:53:10,249 [ERROR   ] retcode: 54
[2019-10-31T22:53:17.380Z]        OK (67.844s)
[2019-10-31T22:54:18.909Z]          test_pam_auth_valid_user (integration.shell.test_auth.AuthTest) ... 16:54:11,486 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:54:18.909Z]        16:54:11,486 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.909Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.909Z]        16:54:11,486 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.909Z]        16:54:11,519 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:54:18.909Z]        16:54:11,520 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.909Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.909Z]        16:54:11,520 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.909Z]        16:54:11,551 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:54:18.909Z]        16:54:11,552 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.909Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.909Z]        16:54:11,552 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.909Z]        16:54:11,582 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltdev']' failed with return code: 54
[2019-10-31T22:54:18.909Z]        16:54:11,583 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.909Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.909Z]        16:54:11,583 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.909Z]        16:54:18,339 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:54:18.910Z]        16:54:18,339 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.910Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.910Z]        16:54:18,340 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.910Z]        16:54:18,373 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:54:18.910Z]        16:54:18,373 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.910Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.910Z]        16:54:18,373 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.910Z]        16:54:18,408 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:54:18.910Z]        16:54:18,409 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.910Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.910Z]        16:54:18,409 [ERROR   ] retcode: 54
[2019-10-31T22:54:18.910Z]        16:54:18,478 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltadm']' failed with return code: 54
[2019-10-31T22:54:18.910Z]        16:54:18,479 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
[2019-10-31T22:54:18.910Z]        <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
[2019-10-31T22:54:18.910Z]        16:54:18,479 [ERROR   ] retcode: 54
[2019-10-31T22:54:26.304Z]        OK (68.509s)```

**Comment by Akm0d** [comment]
Verified, skipping `test_pam_auth_invalid_user` causes py3 mac tests to pass

**Comment by Akm0d** [comment]
This issue is a problem again https://github.com/saltstack/salt-jenkins/issues/504

**Comment by Akm0d** [comment]
Found it! This is the line of code causing all the trouble on mac:
`self.run_call('group.info {0}'.format(self.group))`

**Comment by Akm0d** [comment]
The last PR did it, there might be a problem with the macosxmojave build or with the version of the grp module, but for now I think everything is working :)

**Comment by Akm0d** [comment]
re-run full centos7

**Comment by Akm0d** [comment]
re-run full opensuse15

File: salt/log/handlers/fluent_mod.py
Comment (s0undt3ch) [inline salt/log/handlers/fluent_mod.py]: And do we need this try/except? Doesn't salt.utils.msgpack handle this now?

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Comment by Akm0d** [comment]
re-run full debian10

**Comment by Akm0d** [comment]
re-run full m2crypto

**Comment by Akm0d** [comment]
re-run full centos7

**Comment by Akm0d** [comment]
re-run full amazon2

**Comment by Akm0d** [comment]
Since this is such a large PR we'll break it up into smaller PRs to get it into salt.  The msgpack stuff will come out first.

**Comment by Akm0d** [comment]
This is dependent on https://github.com/saltstack/salt/pull/55354, once that get's merged this one can go in

**Comment by Akm0d** [comment]
```
 test_pam_auth_valid_group (integration.shell.test_auth.AuthTest)
14:44:53         [CPU:0.0%|MEM:40.5%]  ... 14:44:51,449 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
14:44:53           *salt.utils.args.get_function_argspec(original_function)
14:44:53         
14:45:02         14:44:59,692 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
14:45:02           *salt.utils.args.get_function_argspec(original_function)
14:45:02         
14:45:11         14:45:10,718 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
14:45:11           *salt.utils.args.get_function_argspec(original_function)
14:45:11         
14:45:13         14:45:11,639 [WARNING ] /private/tmp/kitchen/testing/salt/modules/file.py:32: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
14:45:13           from collections import Iterable, Mapping, namedtuple
14:45:13         
14:45:13         14:45:12,603 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
14:45:13           *salt.utils.args.get_function_argspec(original_function)
14:45:13         
14:53:29         14:53:21,786 [tests.support.case                                          :271 ][ERROR   ] run_script timed out after 500 seconds (process killed)
14:53:29         14:53:21,786 [tests.support.case                                          :285 ][ERROR   ] run_script results for: salt-call  --config-dir /private/tmp/salt-tests-tmpdir/config group.info saltops
14:53:29         return code: None
14:53:29         stdout:
14:53:29         None
14:53:29         
14:53:29         stderr:
14:53:29         None
14:53:32         14:53:30,395 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
14:53:32           *salt.utils.args.get_function_argspec(original_function)
14:53:32         
14:53:32         14:53:31,219 [WARNING ] /private/tmp/kitchen/testing/salt/modules/file.py:32: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
14:53:32           from collections import Iterable, Mapping, namedtuple
14:53:32         
14:53:32         14:53:32,147 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
14:53:32           *salt.utils.args.get_function_argspec(original_function)
14:53:32         
15:01:51         15:01:41,829 [tests.support.case                                          :271 ][ERROR   ] run_script timed out after 500 seconds (process killed)
15:01:51         15:01:41,829 [tests.support.case                                          :285 ][ERROR   ] run_script results for: salt-call  --config-dir /private/tmp/salt-tests-tmpdir/config group.add saltops
15:01:51         return code: None
15:01:51         stdout:
15:01:51         None
15:01:51         
15:01:51         stderr:
15:01:51         None
15:01:51         15:01:50,797 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
15:01:51           *salt.utils.args.get_function_argspec(original_function)
15:01:51         
15:01:51         15:01:50,833 [WARNING ] /private/tmp/kitchen/testing/salt/modules/file.py:32: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
15:01:51           from collections import Iterable, Mapping, namedtuple
15:01:51         
15:01:51         15:01:50,875 [ERROR   ] Command '['dscl', 'localhost', '-append', '/Local/Default/Groups/saltops', 'GroupMembership', 'saltadm']' failed with return code: 185
15:01:51         15:01:50,876 [ERROR   ] stdout: append: Invalid Path
15:01:51         15:01:50,876 [ERROR   ] stderr: <dscl_cmd> DS Error: -14009 (eDSUnknownNodeName)
15:01:51         15:01:50,876 [ERROR   ] retcode: 185
15:02:16         15:02:13,866 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
15:02:16           *salt.utils.args.get_function_argspec(original_function)
15:02:16         
15:02:16         15:02:13,900 [WARNING ] /private/tmp/kitchen/testing/salt/modules/file.py:32: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
15:02:16           from collections import Iterable, Mapping, namedtuple
15:02:16         
15:02:16         15:02:13,957 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltadm']' failed with return code: 54
15:02:16         15:02:13,958 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:16         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:16         15:02:13,958 [ERROR   ] retcode: 54
15:02:16         15:02:13,992 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltadm']' failed with return code: 54
15:02:16         15:02:13,993 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:16         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:16         15:02:13,993 [ERROR   ] retcode: 54
15:02:16         15:02:14,028 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltadm']' failed with return code: 54
15:02:16         15:02:14,029 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:16         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:16         15:02:14,029 [ERROR   ] retcode: 54
15:02:16         15:02:14,066 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltadm']' failed with return code: 54
15:02:16         15:02:14,066 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:16         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:16         15:02:14,067 [ERROR   ] retcode: 54
15:02:16         FAIL (1051.732s)
15:02:16           test_pam_auth_valid_user (integration.shell.test_auth.AuthTest)
15:02:25         [CPU:0.0%|MEM:42.2%]  ... 15:02:23,907 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
15:02:25           *salt.utils.args.get_function_argspec(original_function)
15:02:25         
15:02:35         15:02:33,892 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
15:02:35           *salt.utils.args.get_function_argspec(original_function)
15:02:35         
15:02:57         15:02:56,976 [WARNING ] /private/tmp/kitchen/testing/salt/utils/decorators/signature.py:31: DeprecationWarning: `formatargspec` is deprecated since Python 3.5. Use `signature` and the `Signature` object directly
15:02:57           *salt.utils.args.get_function_argspec(original_function)
15:02:57         
15:02:57         15:02:57,007 [WARNING ] /private/tmp/kitchen/testing/salt/modules/file.py:32: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
15:02:57           from collections import Iterable, Mapping, namedtuple
15:02:57         
15:02:57         15:02:57,061 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/com.apple.sharepoint.group.1', 'GroupMembership', 'saltdev']' failed with return code: 54
15:02:57         15:02:57,062 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:57         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:57         15:02:57,062 [ERROR   ] retcode: 54
15:02:57         15:02:57,094 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/everyone', 'GroupMembership', 'saltdev']' failed with return code: 54
15:02:57         15:02:57,095 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:57         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:57         15:02:57,095 [ERROR   ] retcode: 54
15:02:57         15:02:57,127 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/_lpoperator', 'GroupMembership', 'saltdev']' failed with return code: 54
15:02:57         15:02:57,127 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:57         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:57         15:02:57,128 [ERROR   ] retcode: 54
15:02:57         15:02:57,159 [ERROR   ] Command '['dscl', 'localhost', '-delete', '/Local/Default/Groups/localaccounts', 'GroupMembership', 'saltdev']' failed with return code: 54
15:02:57         15:02:57,160 [ERROR   ] stderr: <main> attribute status: eDSAttributeNotFound
15:02:57         <dscl_cmd> DS Error: -14134 (eDSAttributeNotFound)
15:02:57         15:02:57,160 [ERROR   ] retcode: 54
15:03:00         OK (43.043s)
```

**Comment by Akm0d** [comment]
re-run full windows2019

**Comment by Akm0d** [comment]
re-run full windows2019

**Comment by Akm0d** [comment]
re-run full ubuntu1604

**Comment by Akm0d** [comment]
Needs testcase based on https://github.com/saltstack/salt/issues/23373

**Comment by Akm0d** [comment]
re-run full macosxmojave

**Comment by Akm0d** [comment]
The command that has issue on mac is `salt -a pam '*' test.ping --username saltadm`.  The error message "Authentication error occurred" is given when an incorrect password is used

**Comment by Akm0d** [comment]
```
[ERROR   ] Command '['dscl', 'localhost', '-append', '/Local/Default/Groups/saltops', 'GroupMembership', 'saltadm']' failed with return code: 185", '13:50:55,820 
[ERROR   ] stdout: append: Invalid Path', '13:50:55,820 
[ERROR   ] stderr: <dscl_cmd> DS Error: -14009 (eDSUnknownNodeName)', '13:50:55,820 
[ERROR   ] retcode: 185']
```

**Comment by Akm0d** [comment]
re-run full amazon2

**Comment by dwoz** [comment]
re-run full all

**Comment by Akm0d** [comment]
The centos7 test failure is a serious issue with this PR:
```Traceback (most recent call last):
  File "/tmp/kitchen/testing/tests/integration/scheduler/test_skip.py", line 301, in test_run_seconds_skip
    self.schedule.eval(now=run_time)
  File "/tmp/kitchen/testing/salt/utils/schedule.py", line 1593, in eval
    now)
  File "/tmp/kitchen/testing/salt/utils/schedule.py", line 224, in _check_max_running
    current_jobs = salt.utils.minion.running(self.opts)
  File "/tmp/kitchen/testing/salt/utils/minion.py", line 33, in running
    data = _read_proc_file(path, opts)
  File "/tmp/kitchen/testing/salt/utils/minion.py", line 73, in _read_proc_file
    data = serial.loads(buf)
  File "/tmp/kitchen/testing/salt/payload.py", line 140, in loads
    exc,
  File "/tmp/kitchen/testing/salt/ext/six.py", line 725, in raise_from
    raise value
SaltDeserializationError: Could not deserialize msgpack message. See log for more info.
```

**Review by dwoz — APPROVED** [review]

**Review by dwoz — DISMISSED** [review]

**Comment by dwoz** [comment]
@Akm0d I believe this is being addressed in a more comprehensive way for the proxy minion work @garethgreenaway is working on. We should hold off on this to see how that work pans out.

**Pull request closed without merging**
